Bug 1807204: Update cursor to show drop status over targets

jeff-phillips-18 opened the pull request:
**Fixes**: 
https://issues.redhat.com/browse/ODC-3028

**Analysis / Root cause**: 
The cursor was not updated during drag operations to indicate drop status.

**Solution Description**: 
Modified the cursor to:
- Use an SVG for the Add cursor when creating connections
- Show a tooltip when over a valid add drop target informing the user that dropping will add an application here (after the standard delay without a mouse move)
- Show an invalid drop cursor when over items where a drop would do nothing

** Sample Screen Shots **
![Drag_Cursor_2](https://user-images.githubusercontent.com/11633780/75288212-d9207180-57e9-11ea-9559-b91c028052b5.gif)

**Browser conformance**: 
- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge

cc @openshift/team-devconsole-ux 

/kind bug

jeff-phillips-18 commented:
/cherry-pick release-4.4

christianvogt commented on frontend/packages/topology/src/behavior/withCreateConnector.tsx:
```diff
@@ -135,6 +135,22 @@ const CreateConnectorWidget: React.FC<CreateConnectorWidgetProps> = observer((pr
+    return () => {
+      clearTimeout(unsetHandle.current);
+    };
+  }, [endPoint.x, endPoint.y]);
```
This is better off going directly into the `DefaultCreateConnector`.

In dev-console we can supply our own prop to `renderConnector` which uses `DefaultCreateConnector` to supply the custom message.

christianvogt commented on frontend/packages/dev-console/src/components/topology/components/nodes/KnativeServiceGroup.tsx:
```diff
@@ -92,7 +92,12 @@ const KnativeServiceGroup: React.FC<KnativeServiceGroupProps> = ({
+      content="Move sink to service"
+      trigger="manual"
+      isVisible={dropTarget && canDrop}
+      tippyProps={{ duration: 0, delay: 0 }}
```
Can you confirm the behavior with UX? 0 delay will bring up tooltips immediately as you hover your mouse over an intermediate node.

Didn't we try avoid this kind of behavior for label truncation by having a small delay.

jeff-phillips-18 replied:
UX spec specifies `For connectors, the tooltip should appear the moment the user hovers over a valid drop target.`

christianvogt commented on frontend/packages/dev-console/src/components/topology/componentUtils.ts:
```diff
@@ -174,9 +174,20 @@ const graphWorkloadDropTargetSpec: DropTargetSpec<
+  collect: (monitor) => {
+    const dragEditInProgress =
+      monitor.isDragging() && editOperations.includes(monitor.getOperation());
+    const dropHints = monitor.getDropHints() || [];
```
maybe we should change `getDropHints` to always return an array (`string[]`)?

jeff-phillips-18 commented:
Updated per comments.

Removed the `grab` cursor when dragging edges over valid drop nodes.

In order to move out the dev-console specific hover treatment for the create connector, I had to refactor a bunch to make the default connector a component rather than a function call that returns a component.

christianvogt commented on frontend/packages/topology/src/components/DefaultCreateConnector.tsx:
```diff
@@ -1,40 +1,66 @@
+import { AddCircleOIcon } from '@patternfly/react-icons';
+import { Tooltip } from '@patternfly/react-core';
 import ConnectorArrow from './ConnectorArrow';
+import { CreateConnectorRenderer } from './CreateConnectorWidget';
```
`CreateConnectorWidget` imports `DefaultCreateConnector`
and `DefaultCreateConnector` imports  `CreateConnectorWidget`
Probably worth combining the two together.

christianvogt commented on frontend/packages/topology/src/components/CreateConnectorWidget.tsx:
```diff
@@ -0,0 +1,206 @@
+import './CreateConnectorWidget.scss';
+
+export const CREATE_CONNECTOR_OPERATION = 'createconnector';
+export const CREATE_CONNECTOR_DROP_TYPE = '#createConnector#';
```
Old code but:

```suggestion
export const CREATE_CONNECTOR_OPERATION = '#createconnector#';
export const CREATE_CONNECTOR_DROP_TYPE = '#createConnector#';
```

christianvogt commented on frontend/packages/topology/src/components/CreateConnectorWidget.tsx:
```diff
@@ -0,0 +1,206 @@
+  className?: string;
+};
+
+export type CreateConnectorRenderer = React.ComponentType<CreateConnectorRendererProps>;
```
This isn't correct. The only required props that are expected of a component used in `withCreateConnector` are:
```
  startPoint: Point,
  endPoint: Point,
  hints: string[] | undefined,
```

christianvogt commented on frontend/packages/topology/src/components/DefaultCreateConnector.tsx:
```diff
@@ -1,40 +1,66 @@
+const cursorSize = 20;
 
-const DefaultCreateConnector: React.FC<DefaultCreateConnectorProps> = ({
+const DefaultCreateConnector: CreateConnectorRenderer = ({
```
You'll need a new prop type that includes support for `tipContents` and `className`.

christianvogt commented on frontend/packages/topology/src/components/index.ts:
```diff
@@ -5,3 +5,5 @@ export { default as ConnectorArrow } from './ConnectorArrow';
 export { default as EdgeConnectorArrow } from './EdgeConnectorArrow';
 export { default as GraphComponent } from './GraphComponent';
 export { default as VisualizationSurface } from './VisualizationSurface';
+export * from './CreateConnectorWidget';
```
Do not expose contents of `CreateConnectorWidget` as API. This is for internal use only to `withCreateConnector`. If needed, re-export any specific types from `withCreateConnector`.

christianvogt commented:
/lgtm
/approve
